geode-core/src/main/java/org/apache/geode/internal/statistics/meters/LegacyStatCounter.java

```diff
@@ -0,0 +1,154 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more contributor license
+ * agreements. See the NOTICE file distributed with this work for additional information regarding
+ * copyright ownership. The ASF licenses this file to You under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package org.apache.geode.internal.statistics.meters;
+
+import io.micrometer.core.instrument.Counter;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.Tag;
+
+import org.apache.geode.Statistics;
+
+/**
+ * Wraps a {@code Counter} to increment and read an associated stat as well as incrementing the
+ * counter.
+ */
+public class LegacyStatCounter implements Counter {
+  private final Counter underlyingCounter;
+  private final StatisticBinding statisticBinding;
+
+  /**
+   * Creates a {@code LegacyStatCounter} that wraps the given counter to add the ability
+   * to increment and read an associated stat.
+   *
+   * @param underlyingCounter the counter to wrap
+   * @param statisticBinding associates the {@code LegacyStatCounter} with a stat
+   */
+  private LegacyStatCounter(Counter underlyingCounter,
+      StatisticBinding statisticBinding) {
+    this.underlyingCounter = underlyingCounter;
+    this.statisticBinding = statisticBinding;
+  }
+
+  /**
+   * Increments both the underlying counter and the associated stat by the given amount.
+   */
+  @Override
+  public void increment(double amount) {
```

Comment:
Geode statistics use both positive and negative amounts in increment calls. Not sure how this translates for meters, but maybe negative is ok.

Reply:
Though Micrometer describes counters as monotonically increasing, nothing in Micrometer enforces that. Given the lack of enforcement in Micrometer, I think I don't want to add that enforcement here.

Reply:
Like Micrometer, Geode distinguished at the API level between counters and gauges but both don't enforce positive increments only for counters. A Micrometer registry may enforce this though so we should not use `Counters` when we want a `Gauge`.

Reply:
Yes. I'm not sure yet what kind of meter we'll use for gauge stats.

Reply:
Well, I would hope the `Gauge` meter. We don't have many so in those cases it may make sense to do what is best. Likely in those cases the service being gauged should probably have method for getting the current observed value and both the Micrometer `Meter` and legacy `Statistics` `LongSupplier` should be setup to poll that method.

Reply:
Many of these might be getting reworked like this already as a result of some of the performance work we are currently working on. Often these gauges were being used to determine behavior of the service it gauged. In the current implementation the service inc/dec some Statistics gauge and then read it when performing some action. This leads to poor performance since the read of `Statistics` is costly. It also leads to a bug when `Statistics` is turned off and the no-op implementation is in use, now all gauges are 0. With Micrometer in the mix the same issues would exist. We are moving all these uses back into the service. The service will keep its own gauge and the `Statistics` will poll it via a `LongSupplier`. The service will not need to poll the value from outside its own objects.

Long story short, we should consider all this and refactor some things when we hit gauges like this doing Micrometer stuff.